### ISSUE TITLE
Pan View high-level design

### ISSUE THREAD

**Issue by xsdg** [BestImageViewer/geeqie#461]
Copying @caclark from #458:
Hello Omari
It's clear from all the work you are doing here that you are a heavy-duty user of Pan View.
I don't use Pan View, but if I did it would be to use the calendar view to see images I took on particular days. This would be far quicker and more visually intuitive than making multiple calls to the search window.

The reason I don't use it is:
a) I can't allocate keywords to a selection
b) I can't allocate a selection to a collection
c) if I use "Show Image", at say 10%, I can't use the mouse to slide the image around so that I can see the underlying thumbnails
d) I can't double-click on a thumbnail to bring up the image in the main window

Do you also experience these shortcomings, or do you use Pan View in a very different way?

**Comment by xsdg** [comment]
@caclark I don't personally use collections, but all of the things you listed are definitely pretty significant shortcomings.  The main thing is that my usecase is handled better by PanView+filtering than it is by anything else in Geeqie.  Here's the kind of thing I do frequently:

1) Take photos at some event,  I'll typically come away with 2,000–4,000 frames.  The example that I've been working on is a 25-hour-long endurance car race.  See here:
http://blog.doppler-photo.net/2016/12/behind-25-journey-to-center-of-race.html
http://blog.doppler-photo.net/2016/12/fractal-nature-behind-25-part-2.html
2) Tag images as `decent`, and then further with things like `morning` or `night`, `racecar`, `car 12`, and so on.  Basically, various keywords with different levels of generality.
3) For each chapter, try to find photos that are relevant to that chapter.  So, for instance, `car 45` doing `pit stop` during `nighttime`.  Or, for instance, cars 5 and 57 were on the same team.  So `car (5|57)`.
4) Of the resulting images, compare them side-by-side-by-side and pick the best one

Fundamentally, step 4 is one major reason why I need Pan View for this.  I can do an okay comparison sometimes by switching back-and-forth between two images.  But with three or more of them, the ability to see them all concurrently is invaluable.

The other major reason I use Pan View is because I can visually jump past large quantities of images with very little effort.  For instance, suppose I'm looking for a particular image and I know "oh, I took that one inside, and it was pretty soon after I took the photo of the starter stand."  Once I find the starter stand picture, I can just zoom out and look for the next picture that's not "bright," and start searching there.  That's much, much, much more efficient to do with the ability to see large numbers of images concurrently.

**Comment by xsdg** [comment]
That said, having built the filtering stuff for Pan View, it's pretty clear that the different feature sets in Pan View versus DirView (and I don't even know about Collections) make for a fairly poor user experience.  I found it incredibly frustrating that I could show images in, for instance, `"car (5|57)" - "on track"`, but then that I couldn't reproduce that particular image set in DirView.

That said, I find the "marks" concept to be pretty straightforward and simple, and it's no good replacing that with a different filtering mechanism that is strictly more complex.  Maybe add the ability to associate marks with filters?  That would be useful for sifting and for filtering, but then it'd be ambiguous what action to take if someone tries to add or remove a mark for a given image.

But again, some kind of MVC pattern seems like it'd make sense here.  You can imagine the model as a generator that just lists files and subdirectories (possibly recursively, possibly recursively-with-max-depth), along with a filter that produces only a subset of those images for whichever views are being used with that model.  You do caching at the model level, so that adding and removing filters is always fast.

Then the views are all siblings and we'd need to build out the right transitions between them.  Double-clicking on an image in Pan View might create a Dir View focusing on that image, as @caclark suggested.

(cc+ @mowgli )

**Comment by xsdg** [comment]
Lastly, Pan View pretty obviously needs a UI overhaul.  We probably can't avoid the custom rendering, given that the UI patterns supported by Pan View don't have really analogs anywhere in GTK.  That said, the Pan View rendering stuff needs to be factored out into a toolkit, and that toolkit needs tests.  There exist obvious layout bugs right now that we should track down, fix, and then verify with tests.  But really, we should defer to GTK rendering to the utmost extent possible, simply because reimplementations of code that already exists is just another place where we can write new bugs that are unique to our codebase.

**Comment by caclark** [comment]
>  I couldn't reproduce that particular image set in DirView

I believe an "image set" should be a Collection - a set of images from various folders presented as a cohesive unit. In #443 I tried to explain - probably not too well - how Collections should work.

Over the years people have occasionally asked for Marks to be preserved. I believe the intention of the designer was for Marks to be a scratchpad (single-session only) as an aid to set Keywords.

The two significant annoyances I have with Geeqie are that I cannot use Pan View/Calendar and Collections as I would wish. And an inconsistent user interface is a few lengths behind in third place. I think some of the Pan View problems can be solved - I'm not sure about Collections.

**Comment by xsdg** [comment]
I agree with you that Marks should not be persistent.  The question was this: currently, I can associate Mark 1 with keyword "fuzzball".  Then, if I set Mark 1 on an image, the keyword "fuzzball" is added.  If I remove Mark 1 from an image, the keyword "fuzzball" is removed.

Suppose I set Mark 2 with a filter "fuzz(ball)?"  Then Mark 2 will show me all of the images that have keywords "fuzz" and/or "fuzzball", but it's ambiguous what it should mean if I try to actively set or remove Mark 2 from an image.  Should an image with both the "fuzz" and "fuzzball" keywords set have them both removed?  The opposite is obviously unachievable — there are an infinite number of strings that a particular regex might match.

One easy solution is just to consider Marks that are driven by filters as read-only.  Dunno if there's something better.

**Comment by caclark** [comment]
I find the marks-keywords facility a bit fraught, except for simple operation - e.g.

If I make a mark-to-keyword link then mark an image, the keyword immediately has a check-mark against it. I then don't know which keyword check-marks relate to already-written metadata, and which to yet-to-be-written metadata (or maybe never written if I cancel out).
If I have several images marked, then decide to link that mark to a keyword, the existing marks get deleted.

There are other problems too.

I think a bit of design refinement would make this an easier to use and perhaps more powerful facility.

**Comment by xsdg** [comment]
Multiple years have passed, so it's the perfect time to revisit this :P 

I think this is a reasonable aspiration (happy to hear other perspectives on this): list view, icon view, and pan view should function as three different points on the continuum of "single-image-focused" to "many-image-focused."  List view makes it easiest to see and work on single images at a time.  Icon view makes it easiest to see and work on ~a dozen images at a time.  Pan view makes it easiest to see and work on many dozens or hundreds of images at a time.

Two obvious sub-goals are to add image selection logic to pan-view, and (as discussed above) to be able to see/set/unset marks from pan-view.

At least at the outset, it'll probably be easiest to keep image selections associated with any particular view.  As also discussed above, it would be independently useful to be able to trivially create a collection from a particular selection, and to use one of the views to show that collection.
